<a href="https://colab.research.google.com/github/Rubait-islam-jami/deep-learning/blob/main/DL_Assignment_03_Answer.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# DL Assignment 03

**Name:*Md.Rubait Islam Jami*

**Course Email:*rubaitjami278@gmail.com*  


## End of Assignment

Before submitting:
- Run all cells from top to bottom.  
- Check that all answer sections are filled.  
- Instruction video অনুযায়ী আমাদের দেয়া Colab ফাইলটি থেকে প্রথম একটি Save copy in drive করে নিবা। এরপর Google colab এর মধ্যে কোডগুলো করবে এবং সেই ফাইলটি ‘Anyone with the link’ & ‘View’ Access দিয়ে ফাইলটির Shareble Link টি সাবমিট করবে।

# General Instruction

You must choose your own dataset.

The dataset must:

Be a supervised learning dataset (Regression or Binary Classification)

Contain at least 300 samples

Have at least 2 input features

Be in CSV format

You are NOT allowed to use Dataset or DataLoader.

You must implement everything manually.

# Question 01: [ Marks 05 ]

## Dataset Preparation

## Using your chosen dataset:

Load the dataset.

Perform necessary preprocessing:

Handle missing values (if any)

Encode categorical variables (if necessary)

Feature scaling (if needed)

Separate features (X) and target (y).

Convert them into NumPy arrays.

Convert them into PyTorch tensors.

Split into training and testing sets.

Clearly explain each preprocessing decision.

# **Write** Answer 01:



In [4]:
import pandas as pd
df=pd.read_csv("/content/train.csv")
df.head()

,battery_power,blue,clock_speed,dual_sim,fc,four_g,int_memory,m_dep,mobile_wt,n_cores,...,px_height,px_width,ram,sc_h,sc_w,talk_time,three_g,touch_screen,wifi,price_range
0,842,0,2.2,0,1,0,7,0.6,188,2,...,20,756,2549,9,7,19,0,0,1,1
1,1021,1,0.5,1,0,1,53,0.7,136,3,...,905,1988,2631,17,3,7,1,1,0,2
2,563,1,0.5,1,2,1,41,0.9,145,5,...,1263,1716,2603,11,2,9,1,1,0,2
3,615,1,2.5,0,0,0,10,0.8,131,6,...,1216,1786,2769,16,8,11,1,0,0,2
4,1821,1,1.2,0,13,1,44,0.6,141,2,...,1208,1212,1411,8,2,15,1,1,0,1


In [5]:
import numpy as np
import torch
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

print(df.isnull().sum())
df=df.dropna()

battery_power    0
blue             0
clock_speed      0
dual_sim         0
fc               0
four_g           0
int_memory       0
m_dep            0
mobile_wt        0
n_cores          0
pc               0
px_height        0
px_width         0
ram              0
sc_h             0
sc_w             0
talk_time        0
three_g          0
touch_screen     0
wifi             0
price_range      0
dtype: int64


In [6]:
X= df.drop("price_range",axis=1)
y=df["price_range"]

X=X.values
y=y.values

scaler=StandardScaler()
X= scaler.fit_transform(X)

X_tensor = torch.tensor(X,dtype=torch.float32)
y_tensor = torch.tensor(y,dtype=torch.long)

X_train,X_test,y_train,y_test=train_test_split(
    X_tensor,y_tensor,test_size=0.2,random_state=42
)

explain:                           
Dataset Loading:                  
Pandas was used to efficiently read and handle structured CSV data.               
Handling Missing Values:    
Missing values were checked to ensure data quality            

Categorical   Encoding:
not applied because are already numerical.        
feature Scaling:             
StandardScaler was used to normalize features so that all are on a simillar use
Feature/Target separation:            
separating  X and Y allows the models to learn the realtionshp betweenminput features and target variables                          

Numpy conversion: data convert to numpy array     
Tensor conversion:for model traing

train-test Split:   
the dataset split 80-20 to evaluate the model on unseen data              



# Question 02: [ Marks 20 ]

## Design a neural network using nn.Module.

### The model must contain:

Input layer

At least one hidden layer

Output layer

Suitable activation function



## Justify:

Number of hidden neurons

Choice of activation function

Print  the total number of trainable parameters.


## Write Answer 02:


In [9]:
import torch.nn as nn

class NeuralNetwork(nn.Module):
    def __init__(self,input_size):
      super().__init__()

      self.fc1 = nn.Linear(input_size,32)
      self.relu = nn.ReLU()

      self.fc2 = nn.Linear(32,4)
    def forward(self,x):
      x=self.relu(self.fc1(x))
      x=self.fc2(x)
      return x
model =NeuralNetwork(X_train.shape[1])

In [10]:
total_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
print("total trainable paraeters:",total_params)

total trainable paraeters: 804


# Question 03: [ Marks 10 ]

Choose an appropriate loss function.

Choose an optimizer.

<br>

Justify your choices based on:

Regression vs Classification

Nature of the dataset

## Write Answer 03:

In [11]:
import torch.nn as nn
import torch.optim as optim

criterion = nn.CrossEntropyLoss()
optimizer =optim.Adam(model.parameters(),lr=0.1)

# Question 04: [ Marks 15 ]

## Implement a full training loop:

Forward pass

Loss computation

Backward pass

Parameter update

Gradient reset

### Requirements:

Train for at least 100 epochs.

Print loss every 10 epochs.

Store training loss history(You can pick your own Data Structure).

Explain clearly what happens in each step of the pipeline.

## Write Answer 04:

In [ ]:
epochs =100
loss_history=[]

for epoch in range(epochs):
  outputs = model(X_train)
  loss= criterion(outputs,y_train)
  loss.backward()

  optimizer.step()
  optimizer.zero_grad()
  loss_history.appned(loss.item())

  if epoch % 10==0 :
    print(f"Epoch {epoch},Loss: {loss.item()}")

explain:                           
Forward Pass: The input data (X_train) is passed through the model to generate predictions   

loss computation: the prediction are compared with ny_train to calculate the error    

Backward Pass: computes gradients indicating how to afjust the weights.     

Parameter Update:             
The optimizer updates the model weights using computed gradients

gradient reset:             
clears previous gradients         Loss :
Epoch is stored and also printed in every 10 epochs



# Question 05: [ Marks 10 ]

## Evaluate the model on test data.

## For regression:

Report MSE and MAE


## For classification:

Report Accuracy

Compare training vs testing performance.

State whether the model is underfitting or overfitting.

## Write Answer 05:

In [14]:
import torch
import torch.nn.functional as F

model.eval()
with torch.no_grad():
  train_outputs = model(X_train)
  test_outputs = model(X_test)

  train_pred = torch.argmax(train_outputs,dim=1)
  test_pred = torch.argmax(test_outputs,dim=1)

  train_acc= (train_pred ==y_train).float().mean().item()
  test_acc= (test_pred ==y_test).float().mean().item()

  print("Train Accuracy: ",train_acc)
  print("Test Accuracy: ",test_acc)



Train Accuracy:  0.2824999988079071
Test Accuracy:  0.2849999964237213


In [15]:
if train_acc > test_acc:
  print("overfitting")
elif train_acc<test_acc:
  print("Underfitting")
else:
  print("good")

Underfitting


# Question 06: [ Marks 20 ]

## Modify at least ONE of the following:

Learning rate

Number of hidden neurons

Number of epochs

### Train again and compare:

Convergence speed

Final performance

Explain how the change affected the model.

## Write Answer 06:

In [17]:
import torch.optim as optim

class ModifyNN(nn.Module):
    def __init__(self,input_size):
      super().__init__()

      self.fc1 = nn.Linear(input_size,64)
      self.relu = nn.ReLU()

      self.fc2 = nn.Linear(64,4)
    def forward(self,x):
      x=self.relu(self.fc1(x))
      x=self.fc2(x)
      return x
model2 =NeuralNetwork(X_train.shape[1])

criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model2.parameters(),lr=0.01)

In [18]:
epochs = 100
loss_history=[]

for epoch in range(epochs):
  outputs = model2(X_train)
  loss= criterion(outputs,y_train)

  loss.backward()
  optimizer.step()
  optimizer.zero_grad()

  loss_history.append(loss.item())
  if epoch % 10==0 :
     print(f"Epoch {epoch},Loss: {loss.item()}")

Epoch 0,Loss: 1.4115415811538696
Epoch 10,Loss: 1.1463507413864136
Epoch 20,Loss: 0.8682968020439148
Epoch 30,Loss: 0.6249758005142212
Epoch 40,Loss: 0.4514670968055725
Epoch 50,Loss: 0.33562901616096497
Epoch 60,Loss: 0.2574809193611145
Epoch 70,Loss: 0.20402662456035614
Epoch 80,Loss: 0.16630679368972778
Epoch 90,Loss: 0.13879095017910004


final:epoch 90-> 0.138

explain:                        
After modifying the learning rate and increasing hidden neurons ,the traing loss decreased smoothly from 1.41 to 0.13 showing stable convergence.The lower learning rate improved stability,while morw neurons increased learing capacity

# Question 07: [ Marks 20 ]


# Training Analysis

Answer the following:

Why must gradients be reset every epoch?

What happens if learning rate is too high?

What happens if learning rate is too small?

Why do we define layers inside the constructor (__init__) and not inside forward()?


## Write Answer 07:

1.Ans:                      
gradient are not automatically cleared.they keep accumulating from previous batches if we do not reset them.then the model will update weights incorrectly.this makes training unstable and gives wrong results

2.Ans:                        
If learning rate is too high the model learns too aggressively.The weight updates becomes too high. Training becomes unstable and can not learn stable patterns

3.Ans:                         
if too small then: the model learns very slowly and takes too many epochs to converge,training becomes inefficient

4.Ans:                            
we dfine in __init__() function because thats where the model structue is created.
The forward() function is only used to describe how data flows through those layers
__init() builds the models once and forward() use it repeatedly